fl4health/model_bases/masked_layers.py

```diff
@@ -0,0 +1,359 @@
+import copy
+from typing import Optional, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch import Tensor
+from torch.nn.common_types import _size_1_t, _size_2_t, _size_3_t
+from torch.nn.modules.utils import _pair, _single, _triple
+from torch.nn.parameter import Parameter
+
+from fl4health.utils.functions import bernoulli_sample
+
+
+class MaskedLinear(nn.Linear):
+    """
+    Implementation of masked linear layers.
+
+    Like regular linear layers (i.e., nn.Linear module), a masked linear layer has a weight and a bias.
+    However, the weight and the bias do not receive gradient in back propagation.
+    Instead, two score tensors - one for the weight and another for the bias - are maintained.
+    In the forward pass, the score tensors are transformed by the Sigmoid function into probability scores,
+    which are then used to produce binary masks via bernoulli sampling.
+    Finally, the binary masks are  applied to the weight and the bias. During training,
+    gradients with respect to the score tensors are computed and used to update the score tensors.
+
+    Args:
+        in_features: size of each input sample
+        out_features: size of each output sample
+        bias: If set to ``False``, the layer will not learn an additive bias.
+            Default: ``True``
+
+    Attributes:
+        weight: weights of the module.
+        bias:  bias of the module.
+        weight_score: learnable scores for the weights. Has the same shape as weight.
+        bias_score: learnable scores for the bias. Has the same shape as bias.
+    """
+
+    __constants__ = ["in_features", "out_features"]
+    in_features: int
+    out_features: int
+    weight: Tensor
+
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        bias: bool = True,
+        device: Optional[torch.device] = None,
+        dtype: Optional[torch.dtype] = None,
+    ) -> None:
+        super().__init__(in_features, out_features, bias, device, dtype)
+        self.in_features = in_features
+        self.out_features = out_features
+        self.weight = Parameter(
+            torch.empty((out_features, in_features), device=device, dtype=dtype), requires_grad=False
```

Comment:
No you are right, this was just an oversight on my part. Using `torch.empty` here is not necessary and would've overridden what happens in `super().__init__()`. I think originally I wanted to modify `self.weight` so that it doesn't receive gradient, but I could've simply reset `weight.requires_grad` instead (this is also how I handled the convolutional layers).